File: examples/cdp-python/cdp_example.py
"""
Drive an Archonum remote browser over CDP (Python + Playwright).

There is no local browser here — Archonum *is* the browser. We connect out to
it over the Chrome DevTools Protocol (CDP), navigate, and take a screenshot.

Two ways to get a CDP endpoint:

  browsers (default) — ask the Archonum API for a per-country browser:
      GET {ARCHONUM_BASE_URL}/api/v1/browsers/?country=..&session=..
        Authorization: Token {ARCHONUM_TOKEN}
      -> endpoint_url
    Needs: ARCHONUM_TOKEN, ARCHONUM_COUNTRY

  gateway — resolve a websocket debugger URL from the CDP gateway:
      GET {ARCHONUM_CDP_BASE_URL}/json/version?username=..&password=..
      -> webSocketDebuggerUrl
    Needs: ARCHONUM_CDP_BASE_URL, ARCHONUM_CDP_USERNAME, ARCHONUM_CDP_PASSWORD

Mode auto-selects from whichever credentials are set; override with CDP_MODE.
"""

import json
import os
import re
import time
import urllib.request
from pathlib import Path
from urllib.parse import urlsplit, urlunsplit, parse_qsl, urlencode

from playwright.sync_api import sync_playwright

DEFAULT_BASE_URL = "https://app.archonum.com"
DEFAULT_CDP_BASE_URL = "http://app.archonum.com:10900"

TARGET_URL = os.environ.get("TARGET_URL", "https://creepjs.org/checker")
OUTPUT_DIR = Path(os.environ.get("OUTPUT_DIR", "out"))
HTTP_TIMEOUT = float(os.environ.get("HTTP_TIMEOUT_MS", "10000")) / 1000
CONNECT_TIMEOUT_MS = float(os.environ.get("CONNECT_TIMEOUT_MS", "30000"))
CONNECT_RETRIES = int(os.environ.get("CONNECT_RETRIES", "5"))
# creepjs computes a fingerprint/trust score client-side, which takes a while.
PAGE_SETTLE_MS = float(os.environ.get("PAGE_SETTLE_MS", "15000"))


def _with_creds(url: str, username: str, password: str) -> str:
    """Append username/password query params to a URL."""
    parts = urlsplit(url)
    query = dict(parse_qsl(parts.query, keep_blank_values=True))
    query["username"] = username
    query["password"] = password
    return urlunsplit(parts._replace(query=urlencode(query)))


def _redact(url: str) -> str:
    """Hide credentials before logging a URL."""
    parts = urlsplit(url)
    query = [(k, "***" if k in ("username", "password") else v)
             for k, v in parse_qsl(parts.query, keep_blank_values=True)]
    return urlunsplit(parts._replace(query=urlencode(query)))


def _get_json(url: str, headers: dict | None = None) -> dict:
    req = urllib.request.Request(url, headers=headers or {})
    with urllib.request.urlopen(req, timeout=HTTP_TIMEOUT) as res:
        return json.load(res)


def resolve_browsers() -> str:
    """browsers mode: ask the Archonum API for a per-country CDP endpoint."""
    token = os.environ.get("ARCHONUM_TOKEN")
    country = (os.environ.get("ARCHONUM_COUNTRY") or "").lower()
    if not token or not country:
        raise SystemExit("set ARCHONUM_TOKEN and ARCHONUM_COUNTRY")
    base = os.environ.get("ARCHONUM_BASE_URL", DEFAULT_BASE_URL).rstrip("/")
    session = os.environ.get("ARCHONUM_SESSION", "example")

    url = f"{base}/api/v1/browsers/?country={country}&session={session}"
    print(f"fetching Archonum browser -> {url}")
    body = _get_json(url, {"Authorization": f"Token {token}"})
    endpoint = body.get("endpoint_url")
    if not endpoint:
        raise SystemExit(f"Archonum returned no CDP endpoint for {country}")
    print(f"  country={body.get('country', country.upper())} {body.get('country_name', '')}")
    return endpoint


def resolve_gateway() -> str:
    """gateway mode: read the ws debugger URL from the CDP gateway."""
    base = os.environ.get("ARCHONUM_CDP_BASE_URL", DEFAULT_CDP_BASE_URL).rstrip("/")
    username = os.environ.get("ARCHONUM_CDP_USERNAME")
    password = os.environ.get("ARCHONUM_CDP_PASSWORD")
    if not username or not password:
        raise SystemExit("set ARCHONUM_CDP_USERNAME and ARCHONUM_CDP_PASSWORD")

    version_url = _with_creds(f"{base}/json/version", username, password)
    print(f"fetching CDP version -> {_redact(version_url)}")
    body = _get_json(version_url)
    ws = body.get("webSocketDebuggerUrl")
    if not ws:
        raise SystemExit("CDP version metadata had no webSocketDebuggerUrl")
    print(f"  Browser={body.get('Browser', '?')}")
    return _with_creds(ws, username, password)


def resolve_endpoint() -> str:
    mode = os.environ.get("CDP_MODE")
    if not mode:
        mode = "browsers" if os.environ.get("ARCHONUM_TOKEN") else "gateway"
    print(f"mode: {mode}")
    return resolve_browsers() if mode == "browsers" else resolve_gateway()


def connect_with_retry(playwright, endpoint: str):
    """connect_over_cdp, retrying transient 502s while the browser spins up."""
    for attempt in range(1, CONNECT_RETRIES + 1):
        try:
            return playwright.chromium.connect_over_cdp(endpoint, timeout=CONNECT_TIMEOUT_MS)
        except Exception as e:  # noqa: BLE001 - retry only on transient gateway errors
            if "502" not in str(e) or attempt == CONNECT_RETRIES:
                raise
            wait = min(2 ** attempt, 8)
            print(f"  connect attempt {attempt} got 502, retrying in {wait}s...")
            time.sleep(wait)


def capture_panel(page, heading: str, clip: dict, out_path: Path) -> dict | None:
    """Find the card containing `heading`, scroll it to the top, screenshot that
    screenful, and pull the numbers out of the card's text. Generic: anchors on
    the leaf element whose own text is the heading, then climbs to the smallest
    ancestor holding the panel's "Total Collectors" summary."""
    handle = page.evaluate_handle(
        """(heading) => {
            const want = heading.toLowerCase();
            const node = [...document.querySelectorAll('*')].find(e =>
                e.childElementCount === 0 &&
                (e.textContent || '').trim().toLowerCase() === want);
            if (!node) return null;
            let el = node;
            while (el.parentElement && !/Total Collectors/i.test(el.textContent || ''))
                el = el.parentElement;
            return el;
        }""",
        heading,
    )
    element = handle.as_element()
    if not element:
        return None
    text = element.inner_text()
    element.evaluate("el => el.scrollIntoView({block: 'start'})")
    page.wait_for_timeout(500)
    page.screenshot(path=str(out_path), animations="disabled", clip=clip)

    def num(label_re: str):
        m = re.search(label_re, text, re.I)
        return m.group(1) if m else None

    return {
        "coverage": num(r"([\d.]+)%\s+Coverage"),
        "successful": num(r"(\d+)\s+Successful"),
        "failed": num(r"(\d+)\s+Failed"),
        "skipped": num(r"(\d+)\s+Skipped"),
        "total_collectors": num(r"(\d+)\s+Total Collectors"),
        "total_time": num(r"([\d.]+ms)\s+Total Time"),
        "avg_per_attempt": num(r"([\d.]+ms)\s+Avg"),
    }


def main() -> None:
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    endpoint = resolve_endpoint()

    with sync_playwright() as playwright:
        print(f"connecting over CDP -> {_redact(endpoint)}")
        browser = connect_with_retry(playwright, endpoint)

        # The remote browser already has a context + page open; reuse them.
        context = browser.contexts[0] if browser.contexts else browser.new_context()
        page = context.pages[0] if context.pages else context.new_page()

        print(f"navigating -> {TARGET_URL}")
        page.goto(TARGET_URL, wait_until="load", timeout=CONNECT_TIMEOUT_MS)
        page.wait_for_timeout(PAGE_SETTLE_MS)

        title = page.title()
        body_text = page.evaluate("() => document.body.innerText.slice(0, 500)")
        print(f"  title -> {title}")
        print(f"  body  -> {body_text.strip()}")

        # The remote browser is a mobile device whose layout viewport is larger
        # than the physical screen, so a default capture tiles the page. Clip to
        # the device screen to get one clean screenful.
        vp = page.evaluate(
            "() => ({iw: innerWidth, ih: innerHeight, sw: screen.width, sh: screen.height})"
        )
        print(f"  viewport -> inner={vp['iw']}x{vp['ih']} screen={vp['sw']}x{vp['sh']}")
        clip = {"x": 0, "y": 0, "width": vp["sw"], "height": vp["sh"]}

        shot = OUTPUT_DIR / "screenshot.png"
        page.screenshot(path=str(shot), animations="disabled", clip=clip)
        print(f"  shot  -> {shot}")

        # On the creepjs checker, grab the "Collector Coverage" summary panel.
        coverage = None
        if "creepjs" in TARGET_URL.lower():
            cov_shot = OUTPUT_DIR / "collector-coverage.png"
            coverage = capture_panel(page, "Collector Coverage", clip, cov_shot)
            if coverage:
                print(f"  coverage -> {coverage}")
                print(f"  panel    -> {cov_shot}")
            else:
                print("  Collector Coverage panel not found")

        result = {
            "endpoint": _redact(endpoint),
            "target": TARGET_URL,
            "url": page.url,
            "title": title,
            "body": body_text.strip(),
            "screenshot": str(shot),
            "collector_coverage": coverage,
        }
        (OUTPUT_DIR / "result.json").write_text(json.dumps(result, indent=2) + "\n")
        print(f"saved -> {OUTPUT_DIR / 'result.json'}")

        browser.close()


if __name__ == "__main__":
    main()

File: examples/cdp-nodejs/cdp-example.js
/**
 * Drive an Archonum remote browser over CDP (Node.js + Playwright).
 *
 * There is no local browser here — Archonum *is* the browser. We connect out
 * to it over the Chrome DevTools Protocol (CDP), navigate, and screenshot.
 *
 * Two ways to get a CDP endpoint:
 *
 *   browsers (default) — ask the Archonum API for a per-country browser:
 *       GET {ARCHONUM_BASE_URL}/api/v1/browsers/?country=..&session=..
 *         Authorization: Token {ARCHONUM_TOKEN}
 *       -> endpoint_url
 *     Needs: ARCHONUM_TOKEN, ARCHONUM_COUNTRY
 *
 *   gateway — resolve a websocket debugger URL from the CDP gateway:
 *       GET {ARCHONUM_CDP_BASE_URL}/json/version?username=..&password=..
 *       -> webSocketDebuggerUrl
 *     Needs: ARCHONUM_CDP_BASE_URL, ARCHONUM_CDP_USERNAME, ARCHONUM_CDP_PASSWORD
 *
 * Mode auto-selects from whichever credentials are set; override with CDP_MODE.
 */

import { mkdir, writeFile } from 'node:fs/promises';
import { join } from 'node:path';
import { chromium } from 'playwright';

const DEFAULT_BASE_URL = 'https://app.archonum.com';
const DEFAULT_CDP_BASE_URL = 'http://app.archonum.com:10900';

const TARGET_URL = process.env.TARGET_URL || 'https://api.ipify.org?format=json';
const OUTPUT_DIR = process.env.OUTPUT_DIR || 'out';
const CONNECT_TIMEOUT_MS = Number(process.env.CONNECT_TIMEOUT_MS) || 30000;
const CONNECT_RETRIES = Number(process.env.CONNECT_RETRIES) || 5;
const PAGE_SETTLE_MS = Number(process.env.PAGE_SETTLE_MS) || 3000;

const withCreds = (url, username, password) => {
  const u = new URL(url);
  u.searchParams.set('username', username);
  u.searchParams.set('password', password);
  return u.toString();
};

const redact = (url) => {
  const u = new URL(url);
  for (const k of ['username', 'password']) if (u.searchParams.has(k)) u.searchParams.set(k, '***');
  return u.toString();
};

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function getJson(url, headers = {}) {
  const res = await fetch(url, { headers });
  if (!res.ok) throw new Error(`HTTP ${res.status} from ${redact(url)}`);
  return res.json();
}

/** browsers mode: ask the Archonum API for a per-country CDP endpoint. */
async function resolveBrowsers() {
  const token = process.env.ARCHONUM_TOKEN;
  const country = (process.env.ARCHONUM_COUNTRY || '').toLowerCase();
  if (!token || !country) throw new Error('set ARCHONUM_TOKEN and ARCHONUM_COUNTRY');
  const base = (process.env.ARCHONUM_BASE_URL || DEFAULT_BASE_URL).replace(/\/$/, '');
  const session = process.env.ARCHONUM_SESSION || 'example';

  const url = `${base}/api/v1/browsers/?country=${country}&session=${session}`;
  console.log(`fetching Archonum browser -> ${url}`);
  const body = await getJson(url, { Authorization: `Token ${token}` });
  if (!body.endpoint_url) throw new Error(`Archonum returned no CDP endpoint for ${country}`);
  console.log(`  country=${body.country || country.toUpperCase()} ${body.country_name || ''}`);
  return body.endpoint_url;
}

/** gateway mode: read the ws debugger URL from the CDP gateway. */
async function resolveGateway() {
  const base = (process.env.ARCHONUM_CDP_BASE_URL || DEFAULT_CDP_BASE_URL).replace(/\/$/, '');
  const username = process.env.ARCHONUM_CDP_USERNAME;
  const password = process.env.ARCHONUM_CDP_PASSWORD;
  if (!username || !password) throw new Error('set ARCHONUM_CDP_USERNAME and ARCHONUM_CDP_PASSWORD');

  const versionUrl = withCreds(`${base}/json/version`, username, password);
  console.log(`fetching CDP version -> ${redact(versionUrl)}`);
  const body = await getJson(versionUrl);
  if (!body.webSocketDebuggerUrl) throw new Error('CDP version metadata had no webSocketDebuggerUrl');
  console.log(`  Browser=${body.Browser || '?'}`);
  return withCreds(body.webSocketDebuggerUrl, username, password);
}

async function resolveEndpoint() {
  const mode = process.env.CDP_MODE || (process.env.ARCHONUM_TOKEN ? 'browsers' : 'gateway');
  console.log(`mode: ${mode}`);
  return mode === 'browsers' ? resolveBrowsers() : resolveGateway();
}

/** connectOverCDP, retrying transient 502s while the browser spins up. */
async function connectWithRetry(endpoint) {
  for (let attempt = 1; attempt <= CONNECT_RETRIES; attempt++) {
    try {
      return await chromium.connectOverCDP(endpoint, { timeout: CONNECT_TIMEOUT_MS });
    } catch (e) {
      if (!String(e).includes('502') || attempt === CONNECT_RETRIES) throw e;
      const wait = Math.min(2 ** attempt, 8);
      console.log(`  connect attempt ${attempt} got 502, retrying in ${wait}s...`);
      await sleep(wait * 1000);
    }
  }
}

async function main() {
  await mkdir(OUTPUT_DIR, { recursive: true });
  const endpoint = await resolveEndpoint();

  console.log(`connecting over CDP -> ${redact(endpoint)}`);
  const browser = await connectWithRetry(endpoint);

  // The remote browser already has a context + page open; reuse them.
  const context = browser.contexts()[0] || (await browser.newContext());
  const page = context.pages()[0] || (await context.newPage());

  console.log(`navigating -> ${TARGET_URL}`);
  await page.goto(TARGET_URL, { waitUntil: 'load', timeout: CONNECT_TIMEOUT_MS });
  await page.waitForTimeout(PAGE_SETTLE_MS);

  const title = await page.title();
  const bodyText = (await page.evaluate(() => document.body.innerText.slice(0, 500))).trim();
  console.log(`  title -> ${title}`);
  console.log(`  body  -> ${bodyText}`);

  // The remote browser is a mobile device whose layout viewport is larger than
  // the physical screen, so a default capture tiles the page. Clip to the
  // device screen to get one clean screenful.
  const vp = await page.evaluate(() => ({ sw: screen.width, sh: screen.height }));
  const clip = { x: 0, y: 0, width: vp.sw, height: vp.sh };

  const shot = join(OUTPUT_DIR, 'screenshot.png');
  await page.screenshot({ path: shot, animations: 'disabled', clip });
  console.log(`  shot  -> ${shot}`);

  const result = {
    endpoint: redact(endpoint),
    target: TARGET_URL,
    url: page.url(),
    title,
    body: bodyText,
    screenshot: shot,
  };
  await writeFile(join(OUTPUT_DIR, 'result.json'), JSON.stringify(result, null, 2) + '\n');
  console.log(`saved -> ${join(OUTPUT_DIR, 'result.json')}`);

  await browser.close();
}

main().catch((e) => {
  console.error(e.message || e);
  process.exit(1);
});
